Data/ByteString/Internal.hs

```diff
@@ -733,7 +735,11 @@ times n (BS fp len)
       memcpy destptr p len
       fillFrom destptr len
   where
-    size = len * fromIntegral n
+    n = toInteger n_raw -- don't mess with lawless Integral instances
+    sizeInteger = toInteger len * n
+    size = if  sizeInteger <= toInteger (maxBound :: Int)
```

Comment:
I'm worried about the cost of `toInteger` / `fromInteger`. Why not ``len <= maxBound `quot` n_raw``?

Reply:
I think for almost all combinations of common platforms and `Integral` types, a single `Int` division will take longer than the conversions to and from `Integer` and the multiplication.

Reply:
`toInteger` / `fromInteger` are likely to allocate things on heap, which outweights even heavy CPU instructions. Division is not that expensive on modern platforms and is much more readable.

Reply:
Hmm... when I look up the typical latencies for 64bit `idiv`, I see things like 42-95 cycles of latency on Coffee Lake, which is large enough that the L3 cache is probably still beating it. I guess "almost all" might be wrong (32bit division in particular seems typically fast enough), but at this point I would want to get some measurements. But I'm not sure what a fair test would be.

Reply:
If you wish to stick with multiplication, please use [`timesWord2#`](https://hackage.haskell.org/package/base-4.14.0.0/docs/GHC-Exts.html#v:timesWord2-35-). Plus [`toIntegralSized`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Bits.html#v:toIntegralSized) to cast `nRaw`.

Reply:
I had forgotten about `toIntegralSized`. But it looks inapplicable here because `stimes` cannot give us a `Bits` constraint.